src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java

```diff
@@ -0,0 +1,279 @@
+import org.springframework.context.annotation.Profile;
+import org.springframework.data.cassandra.core.cql.ReactiveCqlTemplate;
+import org.springframework.data.redis.core.ReactiveStringRedisTemplate;
+import org.springframework.stereotype.Service;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+@Service
+@Slf4j
+@Profile("downsample")
+public class MetricDeletionService {
+
+  private static final String PREFIX_SERIES_SET_HASHES = "seriesSetHashes";
+  private static final String DELIM = "|";
+
+  private final ReactiveCqlTemplate cqlTemplate;
+  private final DataTablesStatements dataTablesStatements;
+  private AppProperties appProperties;
+  private TimeSlotPartitioner timeSlotPartitioner;
+  private DownsampleProperties downsampleProperties;
+  private final ReactiveStringRedisTemplate redisTemplate;
+
+  @Autowired
+  public MetricDeletionService(ReactiveCqlTemplate cqlTemplate,
+      DataTablesStatements dataTablesStatements, AppProperties appProperties,
+      TimeSlotPartitioner timeSlotPartitioner,
+      DownsampleProperties downsampleProperties, ReactiveStringRedisTemplate redisTemplate) {
+    this.cqlTemplate = cqlTemplate;
+    this.dataTablesStatements = dataTablesStatements;
+    this.appProperties = appProperties;
+    this.timeSlotPartitioner = timeSlotPartitioner;
+    this.downsampleProperties = downsampleProperties;
+    this.redisTemplate = redisTemplate;
+  }
+
+  public Mono<Void> deleteMetrics(String tenant, String metricName, List<String> tag,
+      Instant start, Instant end) {
+    if (StringUtils.isAllBlank(metricName)) {
+      return deleteMetricsByTenantId(tenant, start, end);
+    } else if (CollectionUtils.isEmpty(tag)) {
+      return deleteMetricsByMetricName(tenant, metricName, start, end);
+    } else {
+      return deleteMetricsByMetricNameAndTag(tenant, metricName, tag, start, end);
+    }
+  }
+
+  private Mono<Void> deleteMetricsByTenantId(String tenant, Instant start, Instant end) {
+    log.info("inside deleteMetricsByTenantId method with tenant {} ", tenant);
+    return Flux.fromIterable(timeSlotPartitioner.partitionsOverRange(start, end, null))
+        .flatMap(timeSlot -> {
+          return deleteMetricsByTenantId(downsampleProperties.getGranularities(), tenant, timeSlot);
+        }).then();
+  }
+
+  private Mono<Void> deleteMetricsByMetricName(String tenant, String metricName, Instant start,
+      Instant end) {
+    log.info("inside deleteMetricsByMetricName method with tenant {}, metricName {} ", tenant,
+        metricName);
+    return Flux.fromIterable(timeSlotPartitioner.partitionsOverRange(start, end, null))
+        .flatMap(timeSlot -> {
+          Flux<String> seriesSetHashes = getSeriesSetHashFromSeriesSets(tenant,
+              metricName);
+          return deleteMetric(downsampleProperties.getGranularities(), tenant, timeSlot, metricName,
+              seriesSetHashes);
+        }).then();
```

Comment:
Not quite sure why we need these empty `then()`

Reply:
as we are returning void from there instead of its default boolean return value

Reply:
https://www.javaer101.com/en/article/797683.html

I think the use of `.then()` implies further action.